src/traces/scattergl/select.js

```diff
@@ -0,0 +1,70 @@
+/**
+* Copyright 2012-2017, Plotly, Inc.
+* All rights reserved.
+*
+* This source code is licensed under the MIT license found in the
+* LICENSE file in the root directory of this source tree.
+*/
+
+
+'use strict';
+
+var subtypes = require('../scatter/subtypes');
+
+module.exports = function selectPoints(searchInfo, polygon) {
+    var cd = searchInfo.cd,
+        xa = searchInfo.xaxis,
+        ya = searchInfo.yaxis,
+        selection = [],
+        trace = cd[0].trace,
+        curveNumber = trace.index,
+        i,
+        di,
+        x,
+        y;
+
+    var scattergl = cd[0].glTrace;
+    var scene = cd[0].glTrace.scene;
+
+    // TODO: include lines? that would require per-segment line properties
+    var hasOnlyLines = (!subtypes.hasMarkers(trace) && !subtypes.hasText(trace));
+    if(trace.visible !== true || hasOnlyLines) return;
+
+    // filter out points by visible scatter ones
+    // var scatter2d = scattergl.scatter.instance
+
+    if(polygon === false) { // clear selection
+        for(i = 0; i < cd.length; i++) cd[i].dim = 0;
+    }
+    else {
+        for(i = 0; i < cd.length; i++) {
+            di = cd[i];
+            //FIXME: this affects performance for 1e6 points
+            x = xa.c2p(di.x);
+            y = ya.c2p(di.y);
```

Comment:
@dfcreative 

> could you pls elaborate why selecting only visible points won't work

For example, two points have the same exact x/y coordinate, one of them will be rendered on the top, and it'll be duly returned by the point picker (via a bijective rgb mapping) on mouseover, and the occluded point will not be reported - but for the area selector, it's expected that the bottom point is also recovered. This generalizes to arbitrary overlaps esp. that the current hover picker needs to work with the rendered markers in raster space while the area selector needs to check for inclusion of mathematical points (rather than marker pixels) in the lassoed area.